docs/src/writing_good_rules.md

```diff
@@ -35,6 +63,77 @@ Examples being:
 - There is only one derivative being returned, so from the fact that the user called
   `frule`/`rrule` they clearly will want to use that one.
 
+## Structs: constructors and functors
+
+To define an `frule` or `rrule` for a _function_ `foo` we dispatch on the type of `foo`, which is `typeof(foo)`.
+For example, the `rrule` signature would be like:
+
+```julia
+function rrule(::typeof(foo), args...; kwargs...)
+    ...
+    return y, foo_pullback
+end
+```
+
+For a struct `Bar`,
+```julia
+struct Bar
+    a::Float64
+end
+
+(bar::Bar)(x, y) = return bar.a + x + y # functor
+```
+we can define an `frule`/`rrule` for the `Bar` constructor(s), as well as any `Bar` [functors](https://docs.julialang.org/en/v1/manual/methods/#Function-like-objects).
+
+To define an `rrule` for a constructor for a  _type_ `Bar` we need to be careful to dispatch only on `Type{Bar}`.
+For example, the `rrule` signature for a `Bar` constructor would be like:
+```julia
+function ChainRulesCore.rrule(::Type{Bar}, a)
+    ...
+    return Bar(a), Bar_pullback
+end
+```
+
+In particular, be careful not to use `typeof(Bar)` here.
+Because `typeof(Bar)` is `DataType`, using this to define an `rrule`/`frule` will define an `rrule`/`frule` for all constructors.
+
+You can check which to use with `Core.Typeof`:
+
+```julia
+julia> function foo end
+foo (generic function with 0 methods)
+
+julia> typeof(foo)
+typeof(foo)
+
+julia> Core.Typeof(foob)
+typeof(foo)
+
+julia> typeof(Bar)
+DataType
+
+julia> Core.Typeof(Bar)
+Type{Bar}
+
+julia> abstract type AbstractT end
+
+julia> typeof(AbstractT)
+DataType
+
+julia> Core.Typeof(AbstractT)
+Type{AbstractT}
+```
+
+For the functor, use `bar::Bar`, i.e.
+
+```julia
+function ChainRulesCore.rrule(bar::Bar, x, y)
+    ...
```

Comment:
including the code for the pullback i think is important.
Possibly also emphasize above that this is what that first arg is for.
```suggestion
    # Notice the first return is not `NoTangent()`
    Bar_pullback(z̄) = Tangent{Bar}(;a=z̄), z̄, z̄
```